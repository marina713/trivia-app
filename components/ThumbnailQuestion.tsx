import React from 'react';
import { View, StyleSheet } from 'react-native';
import { decode } from 'html-entities';
import { Text } from './Themed';
import { Item } from "../types";


const ThumbnailQuestion = ({ item }: { item: Item }) => {
  const color = item.correct_answer === item.userAnswer ? "#90be6d" : "#e76f51";
  return (
    <View style={{ ...styles.container, borderColor: color }}>
      <View style={styles.questionContainer} >
        <Text style={styles.question}>
          {decode(item.question)}
        </Text>
      </View>
      <View style={{ ...styles.buttonContainer, backgroundColor: color }} >
        <Text style={styles.buttonText}> {item.userAnswer}</Text>
      </View>
    </View>)
}

const styles = StyleSheet.create({
  container: {
    marginTop: 7,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    padding: 10,
    borderRadius: 10,
    borderWidth: 1.5
  },
  buttonContainer: {
    paddingVertical: 5,
    width: 57,
    borderRadius: 8,
    justifyContent: 'center',
    alignItems: 'center',
  },
  buttonText: {
    fontSize: 12,
    color: 'white',
  },
  question: {
    fontSize: 13,
  },
  questionContainer: {
    width: '80%',
  }
});

export default ThumbnailQuestion;