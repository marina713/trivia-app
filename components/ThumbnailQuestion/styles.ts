import styled from "styled-components/native";
import { Text } from "../Themed";

export const Container = styled.View<any>`
  margin-top: 7px;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  padding: 10px;
  border-radius: 10px;
  border-width: 1.5px;
  border-color: ${(props) => props.borderColor};
`;

export const QuestionContainer = styled.View`
  width: 80%;
`;
export const ButtonContainer = styled.View<any>`
  padding: 0 5px;
  width: 57px;
  border-radius: 8px;
  justify-content: center;
  align-items: center;
  background-color: ${(props) => props.backgroundColor};
  height: 27px;
`;

export const ButtonText = styled(Text)`
  font-size: 12px;
  color: white;
`;

export const QuestionText = styled(Text)`
  font-size: 13px;
`;
