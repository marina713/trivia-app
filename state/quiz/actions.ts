import {
  SUBMIT_ANSWER,
  RESET,
  REQUEST_DATA,
  REQUEST_DATA_SUCCESS,
  RequestDataPayload,
} from "./constants";
import { Data } from "../../types";

export const submitAnswer = (value: string, index: number) => {
  return {
    type: SUBMIT_ANSWER,
    payload: { value, index },
  };
};

export const reset = () => {
  return {
    type: RESET,
  };
};

export const requestData = (payload: RequestDataPayload) => {
  return {
    type: REQUEST_DATA,
    payload,
  };
};

export const requestDataSuccess: any = (value: Data) => {
  return {
    type: REQUEST_DATA_SUCCESS,
    payload: { value },
  };
};
