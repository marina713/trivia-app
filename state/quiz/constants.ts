import type { Data, Item } from "../../types";

export type StateProps = {
  data: Data;
  currentItem: Item | null;
  currentIdx: number;
};

export type RequestDataPayload = {
  onSuccess: () => void;
  onFail: () => void;
};

export const initialState: StateProps = {
  data: [],
  currentItem: null,
  currentIdx: 0,
};

export const SUBMIT_ANSWER = "quiz/SUBMIT_ANSWER";
export const RESET = "quiz/RESET";
export const REQUEST_DATA = "REQUEST_DATA";
export const REQUEST_DATA_SUCCESS = "REQUEST_DATA_SUCCESS";
