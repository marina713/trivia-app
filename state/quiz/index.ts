import {
  initialState,
  SET_DATA,
  SET_CURRENT_ITEM,
  SUBMIT_ANSWER,
  RESET,
} from "./constants";

const quizReducer = (state = initialState, action: any) => {
  switch (action.type) {
    case SET_DATA: {
      const {
        payload: { value },
      } = action;
      return {
        ...state,
        data: value,
      };
    }

    case SET_CURRENT_ITEM: {
      const {
        payload: { value },
      } = action;
      return {
        ...state,
        currentItem: value,
      };
    }
    case SUBMIT_ANSWER: {
      const {
        payload: { value, index },
      } = action;
      const updatedValue = {
        ...state.data[index],
        userAnswer: value,
      };
      return {
        ...state,
        currentIdx: index + 1,
        data: state.data.map((item, i) => (index === i ? updatedValue : item)),
      };
    }
    case RESET: {
      return initialState;
    }

    default:
      return state;
  }
};

export default quizReducer;
