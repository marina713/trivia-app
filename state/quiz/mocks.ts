export const mockData = [
  {
    category: "Entertainment: Video Games",
    type: "boolean",
    difficulty: "hard",
    question: "Is Spiderman faster than Batman?",
    correct_answer: "True",
    incorrect_answers: ["False"],
    userAnswer: "True",
  },
  {
    category: "Entertainment: Video Games",
    type: "boolean",
    difficulty: "hard",
    question: "Is Hulk faster than Batman?",
    correct_answer: "False",
    incorrect_answers: ["True"],
    userAnswer: "True",
  },
  {
    category: "Entertainment: Video Games",
    correct_answer: "True",
    difficulty: "hard",
    incorrect_answers: ["False"],
    question:
      "In &quot;The Sims&quot; series, the most members in a household you can have is 8.",
    type: "boolean",
  },
];
export const mockState = {
  quiz: {
    data: mockData,
    currentIdx: 2,
  },
};
