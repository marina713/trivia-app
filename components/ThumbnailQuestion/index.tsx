import React from 'react';
import { decode } from 'html-entities';
import { Item } from "../../types";
import { Container, QuestionContainer, ButtonContainer, ButtonText, QuestionText } from './styles';


const ThumbnailQuestion = ({ item }: { item: Item }) => {
  const color = item.correct_answer === item.userAnswer ? "#90be6d" : "#e76f51";
  return (
    <Container borderColor={color}>
      <QuestionContainer >
        <QuestionText>
          {decode(item.question)}
        </QuestionText>
      </QuestionContainer>
      <ButtonContainer backgroundColor={color} >
        <ButtonText>
          {item.userAnswer}
        </ButtonText>
      </ButtonContainer>
    </Container>)
}

export default ThumbnailQuestion;