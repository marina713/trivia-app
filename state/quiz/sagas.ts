import { call, put, takeEvery, Effect } from "redux-saga/effects";
import { REQUEST_DATA, RequestDataAction } from "./constants";
import { requestDataSuccess } from "./actions";

const getQuizData = async () => {
  const response = await fetch(
    "https://opentdb.com/api.php?amount=10&difficulty=hard&type=boolean"
  );
  const json = await response.json();
  return json.results;
};

// worker Saga: will be fired on REQUEST_DATA actions
function* requestData(action: RequestDataAction): Generator<Effect, any, any> {
  try {
    const data = yield call(getQuizData);
    yield put(requestDataSuccess(data));
    yield call(action.payload.onSuccess);
  } catch (e: any) {
    yield put({ type: "DATA_FETCH_FAILED", message: e.message });
    yield call(action.payload.onFail);
  }
}

/*
  Starts requestData on each dispatched `REQUEST_DATA` action.
*/
function* rootSaga() {
  yield takeEvery(REQUEST_DATA, requestData);
}

export default rootSaga;
