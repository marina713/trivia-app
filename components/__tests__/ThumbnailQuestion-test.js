import * as React from "react";
import renderer from "react-test-renderer";

import ThumbnailQuestion from "../ThumbnailQuestion";
import { mockData } from "../../state/quiz/mocks";

it(`renders correctly`, () => {
  const tree = renderer
    .create(<ThumbnailQuestion item={mockData[0]} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
