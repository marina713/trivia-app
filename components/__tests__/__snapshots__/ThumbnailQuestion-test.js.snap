// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  borderColor="#90be6d"
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderColor": "#90be6d",
        "borderRadius": 10,
        "borderWidth": 1.5,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginTop": 7,
        "paddingBottom": 10,
        "paddingLeft": 10,
        "paddingRight": 10,
        "paddingTop": 10,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "width": "80%",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000",
          },
          Array [
            Object {
              "fontSize": 13,
            },
          ],
        ]
      }
    >
      Is Spiderman faster than Batman?
    </Text>
  </View>
  <View
    backgroundColor="#90be6d"
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#90be6d",
          "borderRadius": 8,
          "height": 27,
          "justifyContent": "center",
          "paddingBottom": 0,
          "paddingLeft": 5,
          "paddingRight": 5,
          "paddingTop": 0,
          "width": 57,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000",
          },
          Array [
            Object {
              "color": "white",
              "fontSize": 12,
            },
          ],
        ]
      }
    >
      True
    </Text>
  </View>
</View>
`;
